File: r/lib/positions.R
ensure_player_positions <- function(season, pos_path) {
  if (is.na(pos_path) || pos_path == "") {
    return(invisible(TRUE))
  }
  if (file.exists(pos_path)) {
    return(invisible(TRUE))
  }

  suppressPackageStartupMessages({
    library(dplyr)
    library(stringr)
    library(nflfastR)
  })

  ro_raw <- nflfastR::fast_scraper_roster(seasons = as.integer(season))

  # In your environment, this roster provides gsis_id (00-00xxxxxx), not player_id UUIDs.
  if (!("gsis_id" %in% names(ro_raw))) {
    stop(
      "fast_scraper_roster() did not return gsis_id. Columns found: ",
      paste(names(ro_raw), collapse = ", ")
    )
  }

  ro <- ro_raw %>%
    transmute(
      player_id = as.character(gsis_id), # IMPORTANT: write player_id = GSIS id
      position = toupper(str_trim(position))
    ) %>%
    filter(!is.na(player_id), player_id != "") %>%
    mutate(
      position = str_replace(position, "^FB$", "RB"),
      position_bucket = case_when(
        position %in% c("QB", "RB", "WR", "TE", "K") ~ position,
        TRUE ~ "OTH"
      )
    ) %>%
    distinct(player_id, position_bucket)

  write.csv(ro, pos_path, row.names = FALSE)
  invisible(TRUE)
}

File: src/pbp/positions.py
from __future__ import annotations

from pathlib import Path

import pandas as pd
import nflreadpy as nfl


def _to_pandas(df) -> pd.DataFrame:
    """
    nflreadpy returns Polars DataFrames; convert safely to pandas.
    """
    if df is None:
        return pd.DataFrame()
    if isinstance(df, pd.DataFrame):
        return df
    to_pandas = getattr(df, "to_pandas", None)
    if callable(to_pandas):
        return to_pandas()
    return pd.DataFrame(df)


def ensure_player_positions(season: int, pos_path: str | Path | None) -> None:
    """
    Replacement for the old nfl_data_py-based implementation.

    Writes player_positions_{season}.csv if missing.

    IMPORTANT: uses GSIS id as player_id to match your existing environment
    (mirrors your R logic using gsis_id). :contentReference[oaicite:0]{index=0}
    """
    if not pos_path:
        return

    pos_path = Path(pos_path)
    if pos_path.exists():
        return

    ro_raw_pl = nfl.load_rosters(seasons=season)
    ro_raw = _to_pandas(ro_raw_pl)
    if ro_raw.empty:
        raise RuntimeError(f"load_rosters(seasons={season}) returned no rows")

    # Your R code expects gsis_id. :contentReference[oaicite:1]{index=1}
    if "gsis_id" not in ro_raw.columns:
        raise RuntimeError(
            "load_rosters() did not return gsis_id. Columns found: "
            + ", ".join(map(str, ro_raw.columns))
        )

    ro = pd.DataFrame(
        {
            "player_id": ro_raw["gsis_id"].astype("string"),
            "position": ro_raw.get("position").astype("string"),
        }
    )

    ro["position"] = ro["position"].str.strip().str.upper()
    ro = ro.loc[ro["player_id"].notna() & (ro["player_id"] != "")].copy()

    # Match your R behavior: FB -> RB
    ro["position"] = ro["position"].str.replace(r"^FB$", "RB", regex=True)

    def bucket(pos: str) -> str:
        if pos in {"QB", "RB", "WR", "TE", "K"}:
            return pos
        return "OTH"

    ro["position_bucket"] = ro["position"].fillna("").map(bucket).astype("string")
    ro = ro[["player_id", "position_bucket"]].drop_duplicates()

    pos_path.parent.mkdir(parents=True, exist_ok=True)
    ro.to_csv(pos_path, index=False)
